### PR TITLE
[HOPSWORKS-3230][FSTORE-330] Single expectation api

### PR DESCRIPTION
**Pull request by vatj**
This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/feature_group.py [hunk 1]
```diff
@@ -742,6 +769,20 @@ def __init__(
 
             self.statistics_config = statistics_config
             self.expectation_suite = expectation_suite
+            if expectation_suite:
+                self._expectation_suite._init_expectation_engine(
+                    feature_store_id=featurestore_id, feature_group_id=self._id
+                )
+            self._expectation_suite_engine = (
```
Comment (moritzmeister) [inline python/hsfs/feature_group.py:776]: These should always be set up?

Comment (moritzmeister) [reply python/hsfs/feature_group.py:776]: I suppose this is one of the parts you want to talk about.

File: python/hsfs/feature_group.py [hunk 2]
```diff
@@ -1072,8 +1107,9 @@ def insert(
                 * key `run_validation` boolean value, set to `False` to skip validation temporarily on ingestion.
                 * key `save_report` boolean value, set to `False` to skip upload of the validation report to Hopsworks.
                 * key `ge_validate_kwargs` a dictionary containing kwargs for the validate method of Great Expectations.
+
         # Returns
-            `FeatureGroup`. Updated feature group metadata object.
+            A tuple with job information if python engine is used and the validation report if validation is enabled.
```
Comment (moritzmeister) [inline python/hsfs/feature_group.py:1112]: maybe put 
```
(`Job`, `ValidationReport`)
```
at the beginning

File: python/hsfs/expectation_suite.py [hunk 3]
```diff
@@ -48,18 +54,39 @@ def __init__(
...
-        self._featuregroup_id = featuregroup_id
         self._ge_cloud_id = ge_cloud_id
         self._data_asset_type = data_asset_type
         self._run_validation = run_validation
         self._validation_ingestion_policy = validation_ingestion_policy.upper()
         self._expectations = []
+        self._href = href
+
+        if href is not None:
+            self._init_feature_store_and_feature_group_ids_from_href(href)
```
Comment (moritzmeister) [inline python/hsfs/expectation_suite.py:65]: I don't understand why this is necessary, when you initialize all the objects, iterating through the "items" of the RestDTO, they should all have the feature store and group id set?

File: python/hsfs/expectation_suite.py [hunk 4]
```diff
@@ -128,37 +159,267 @@ def to_ge_type(self):
...
+                    feature_group_id=self._feature_group_id,
+                )
+            )
+        else:
+            raise ValueError(
+                "Provide feature_store_id or feature_group_id to use the expectation suite API"
+            )
+
+    # Expectation Suite Convenience Operations
+    def save(
```
Comment (moritzmeister) [inline python/hsfs/expectation_suite.py:205]: These kind of methods we usually keep in the "engine" classes, e.g. `ExpectationSuiteEngine`

File: python/hsfs/core/great_expectation_engine.py [hunk 5]
```diff
@@ -37,7 +66,12 @@ def validate(self, feature_group, dataframe, save_report, validation_options):
 
                 save_report = validation_options.get("save_report", save_report)
                 if save_report:
-                    return feature_group.save_validation_report(report, ge_type=False)
+                    return feature_group.save_validation_report(report, ge_type=ge_type)
 
-                return validation_report.ValidationReport(**report.to_json_dict())
+                if ge_type:
```
Comment (moritzmeister) [inline python/hsfs/core/great_expectation_engine.py:71]: What's the use case for this?
Not saving the report but getting it as GE?

File: python/hsfs/core/great_expectation_engine.py [hunk 6]
```diff
@@ -15,15 +15,44 @@
...
+
+        :param feature_store_id: id of the respective Feature Store
+        :type `int`
+        :return: The engine for Great Expectation objects
+        :rtype: `GreatExpectationEngine`
+        """
         self._feature_store_id = feature_store_id
 
-    def validate(self, feature_group, dataframe, save_report, validation_options):
+    def validate(
```
Comment (moritzmeister) [inline python/hsfs/core/great_expectation_engine.py:34]: This function needs unit tests

File: python/hsfs/expectation_suite.py (hunk 4 above)
Comment (vatj) [reply python/hsfs/expectation_suite.py:205]: Yes, I am a bit worried people use `expectation_suite.expectations.append(my_expectation)` and then want to persist it. I will remove it

File: python/hsfs/expectation_suite.py [hunk 7]
```diff
@@ -128,37 +159,228 @@ def to_ge_type(self):
...
+            if ge_type:
+                return converted_expectation.to_ge_type()
+            else:
+                return converted_expectation
+        else:
+            raise FeatureStoreException(
+                "Initialize Expectation Suite by attaching to a Feature Group to enable single expectation API"
+            )
+
+    def remove_expectation(self, expectation_id: Optional[int] = None) -> None:
```
Comment (moritzmeister) [inline python/hsfs/expectation_suite.py:332]: example how to get id

File: python/hsfs/expectation_suite.py (hunk 3 above)
Comment (moritzmeister) [reply python/hsfs/expectation_suite.py:65]: let's follow up in another Jira

**Review by moritzmeister — APPROVED** [review]

**Review by moritzmeister — APPROVED** [review]

**Pull request merged**